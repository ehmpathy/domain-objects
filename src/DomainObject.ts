import { Schema } from 'joi';
import { validate } from './validate/validate';

/**
 * Domain Object
 *
 * Responsibilities:
 * - optionally validate the properties that are passed into the constructor against the schema at runtime, if schema is supplied
 * - assign all properties that are passed into constructor to self, after optional runtime validation
 */
export class DomainObject<T> {
  constructor(props: T) {
    // 1. validate with the schema if provided
    const { schema } = this.constructor as typeof DomainObject; // `this.constructor` does not get typed to DomainObject automatically by ts; https://stackoverflow.com/questions/33387318/access-to-static-properties-via-this-constructor-in-typescript
    if (schema) validate({ props, schema, domainObjectName: this.constructor.name });

    // 2. assign all properties to self if passed validation
    Object.assign(this, props);
  }

  /**
   * DomainObject.schema
   *
   * When set, will be used to validate the properties passed into the constructor at runtime (i.e., during instantiation)
   */
  public static schema?: Schema;
}
