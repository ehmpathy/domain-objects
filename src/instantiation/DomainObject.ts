import { withImmute, WithImmute } from '../manipulation/immute/withImmute';
import { DomainObjectShape } from '../reference/DomainReferenceable';
import { hydrateNestedDomainObjects } from './hydrate/hydrateNestedDomainObjects';
import { SchemaOptions, validate } from './validate/validate';

export interface DomainObjectInstantiationOptions {
  /**
   * allow callers to skip certain aspects of instantiation
   *
   * e.g., for performance optimizations in specific usecases
   */
  skip?: {
    /**
     * allow callers to skip schema validation
     *
     * usecase examples
     * - deserialize: schema validation increases deserialization time dramatically (10-100x) (especially if using Joi, that thing is slow!)
     */
    schema?: boolean;
  };
}

/**
 * Domain Object
 *
 * Responsibilities:
 * - optionally validate the properties that are passed into the constructor against the schema at runtime, if schema is supplied
 * - assign all properties that are passed into constructor to self, after optional runtime validation
 */
export class DomainObject<T extends DomainObjectShape> {
  constructor(props: T, options?: DomainObjectInstantiationOptions) {
    // 1. validate with the schema if provided
    const { schema } = this.constructor as typeof DomainObject; // `this.constructor` does not get typed to DomainObject automatically by ts; https://stackoverflow.com/questions/33387318/access-to-static-properties-via-this-constructor-in-typescript
    if (schema && !(options?.skip?.schema === true))
      validate({ props, schema, domainObjectName: this.constructor.name });

    // 2. hydrate any nested props present; just overwrite the orig props for each "nested" key
    const nested = ((this.constructor as typeof DomainObject).nested ??
      {}) as Record<string, typeof DomainObject>;
    const hydratedProps = hydrateNestedDomainObjects({
      domainObjectName: this.constructor.name,
      props,
      nested,
    });

    // 3. assign all properties to self if passed validation
    Object.assign(this, hydratedProps);
  }

  /**
   * DomainObject.alias
   *
   * When set, declares the alias via which domain objects of this class are known to be referred to.
   *
   * For example
   * - PlantPot.alias = 'pot'
   * - Bird.alias = { singular: 'bird', plural: 'flock' }
   *
   * Relevance
   * - formalizes that we can expect to see the dobj referred to by this alias
   * - enables code generators to use this colloquial alias in contracts for better devex
   */
  public static alias?: string | { plural?: string; singular?: string };

  /**
   * DomainObject.schema
   *
   * When set, will be used to validate the properties passed into the constructor at runtime (i.e., during instantiation)
   *
   * Supports [`Zod`](https://github.com/colinhacks/zod), [`Joi`](https://github.com/sideway/joi), and [`Yup`](https://github.com/jquense/yup)
   */
  public static schema?: SchemaOptions<any>; // todo: work around typescript's "static members cant reference class type parameters"

  /**
   * DomainObject.nested
   *
   * When set, will be used to hydrate nested DomainObjects. This is especially useful when instantiating nested domain objects from data stores (e.g., apis or databases). This is _required_ in order to safely use the `serialize` and `getUniqueIdentifier` functions - as those functions check static properties of domain objects to function correctly.
   *
   * For example:
   * ```ts
   * // define the plant pot
   * interface PlantPot {
   *   diameterInInches: number;
   * }
   * class PlantPot extends DomainObject<PlantPot> implements PlantPot {}
   *
   * // define the plant owner
   * interface PlantOwner {
   *   diameterInInches: number;
   * }
   * class PlantOwner extends DomainObject<PlantOwner> implements PlantOwner {}
   *
   * // define the plant
   * interface Plant {
   *   pot: PlantPot;
   *   owners: PlantOwner[];
   *   lastWatered: string;
   * }
   * class Plant extends DomainObject<Plant> implements Plant {
   *   public static nested = { pot: PlantPot, owners: PlantOwner }
   * }
   *
   * // and show that it hydrates the plant pot
   * const plant = new Plant({ pot: { diameterInInches: 7 }, owners: [{ name: 'bob' }], lastWatered: 'monday' }); // notice how the param of `pot` we're inputting is _not_ an instance of `PlantPot`, and `owners` are not instances of `PlantOwner`
   * expect(plant.pot).toBeInstanceOf(PlantPot); // now succeeds, as during instantiation we hydrated `pot` into an instance of `PlantPot`, due to `Plant.nested.pot` being set
   * plant.owners.forEach((owner) => expect(owner).toBeInstance(PlantOwner)); // also succeeds, since during instantiation we hydrated each `owner` into an instance of `PlantOwner`, due to `Plant.nested.owners` being set
   * ```
   */
  public static nested?: Record<string, DomainObject<any>>; // TODO: find a way to make this Record<keyof string>

  /**
   * DomainObject.metadata
   *
   * When set, customizes the keys that are considered as metadata of this domain object.
   *
   * Context,
   * - domain objects are often persisted inside of storage mechanisms that assign metadata to them, such as ids or timestamps
   * - metadata simply adds information _about_ the object, without contributing to _defining_ the object
   *
   * Relevance,
   * - metadata properties do not contribute to the unique key of a DomainLiteral
   * - metadata properties can be easily stripped from an object by using the `omitMetadataValues` method
   *
   * By default,
   * - `id`, `createdAt`, `updatedAt`, and `effectiveAt` are considered metadata keys
   * - `uuid` is also considered a metadata key, if it is not included in the unique key of the DomainEntity or DomainEvent
   */
  public static metadata: readonly string[];

  /**
   * .what = an interface via which to construct instances w/ immute operations
   * .why =
   *   - immute operations such as .clone produce more maintainable code by preventing unexpected mutations
   *   - these immute operations provide a safe pit of success for common operations
   * .note =
   *   - you can add withImmute to any dobj yourself, even if it wasn't built via this .build procedure
   *   - you can override the .build to add your own domain's getters, too
   */
  static build<TInstance extends DomainObjectShape>(
    this: new (props: TInstance) => TInstance,
    props: ConstructorParameters<typeof this>[0],
  ): WithImmute<TInstance> {
    const instance = new this(props);
    return withImmute(instance);
  }
}
