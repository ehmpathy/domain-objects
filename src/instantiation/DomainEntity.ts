import { DomainObject } from './DomainObject';

/**
 * In Domain Driven Design, an Entity is a type of Domain Object for which:
 * - properties change over time (e.g., it has a life cycle)
 * - identity matters (e.g., two entities could have the same properties, differing only by id, and still be valid)
 *
 * For example,
 * - A `User { uuid, name, address, profilePicture }` is an entity. Properties like `name`, `address`, and `profilePicture` can change over time.
 * - A `Job { uuid, requestId, status }` is an entity. We expect `status` to change over time as the job goes from, for example, `QUEUED` to `FULFILLED`.
 */
export class DomainEntity<T> extends DomainObject<T> {
  /**
   * `DomainEntity.unique` defines all of the properties of the entity that the entity is naturally unique on
   *
   * For example,
   * - a `FileProcessingJob { uuid, filePath, status }` is likely going to be unique on the `filePath` if we only ever need to process the same file once.
   * - a `SupportTicket { uuid, userId, type, status }`, on the other hand, is likely only going to be unique on the `uuid`, because the same user can create the same support ticket over and over again
   */
  public static unique: readonly string[];

  /**
   * `DomainEntity.updatable` defines all of the properties of the entity that are updatable
   *
   * For example,
   * - a `User { uuid, name, address, picture }` would likely have `unique = ['name', 'address', 'picture']`, to enable the user to update their information as it changes over time
   * - an `Email { uuid, externalId, toAddress, fromAddress, message, status }` is likely only going to have `status` be updatable, as by changing the `toAddress` would, in this case, make us consider it to be a different email!
   */
  public static updatable: readonly string[];
}
