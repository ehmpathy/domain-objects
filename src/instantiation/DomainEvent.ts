import { DomainObject } from './DomainObject';

/**
 * In Domain Driven Design, an Event is a type of Domain Object for which:
 * - properties are immutable (e.g., something happened)
 * - identity does matter
 *   - i.e., it is uniquely identified by what happened and to what
 *   - i.e., if you change the uniquely identifying properties (when or to what) of an event, it will be considered a new event
 *   - i.e., not all properties on the event uniquely identify the event
 *
 * The purpose of a Domain Event is to represent interesting things that happen in the domain.
 *
 * For example,
 * - An `ItemViewedEvent { itemUuid, customerUuid, occurredAt }` is an event. Changing when it `occurredAt` which `customerUuid` did it or which `itemUuid` it happened to would result in a new event.
 * - A `PowerSupplyMeasuredEvent { batteryUuid, current, voltage, temperature, occurredAt }` is an event. Changing when it `occurredAt` or which `batteryUuid` it occurred to would result in a new event. Note: changing the `current`, `voltage`, or `temperature` would _not_ result in a new event though, since only one measurement can be made at a time for a `batteryUuid`
 */
export abstract class DomainEvent<
  T extends Record<string, any>,
> extends DomainObject<T> {
  /**
   * `DomainEvent.unique` defines all of the properties of the event that the event is naturally unique on.
   *
   * It should always include a timestamp.
   *
   * For example,
   * - a `FileOpenedEvent { fileUuid, userUuid, occurredAt }` is likely going to be unique on the all of the keys: the combination of `['fileUuid', 'userUuid', 'occurredAt']` uniquely identifies the event.
   * - an `AirComfortMeasuredEvent { locationUuid, sensorUuid, temperature, humidity, occurredAt }`, on the other hand, is likely only going to be unique on `['locationUuid', 'sensorUuid', 'occurredAt']`, because logically we cant have a measurement from the same `sensor` + `location` at the same `time`
   */
  public static unique: readonly string[];
}
