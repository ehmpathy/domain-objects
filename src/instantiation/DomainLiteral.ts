import { DomainObject } from './DomainObject';

/**
 * In Domain Driven Design, a Literal (a.k.a. Value Object), is a type of Domain Object for which:
 * - properties are immutable
 *   - i.e., it represents some literal value which happens to have a structured object shape
 *   - i.e., if you change the value of any of its properties, it is a different literal
 * - identity does not matter
 *   - i.e., it is uniquely identifiable by its non-metadata properties
 *
 * The purpose of a Domain Literal is to represent distinct, literal values which happen to have a structured object shape
 *
 * For example,
 * - A `Address { street, city, state, country }` is a literal. Changing any of the properties, like `street`, produces a completely new address
 * - A `Geocode { latitude, longitude }` is a literal. Changing either property means you are dealing with a new geocode
 */
export abstract class DomainLiteral<
  T extends Record<string, any>,
> extends DomainObject<T> {}
