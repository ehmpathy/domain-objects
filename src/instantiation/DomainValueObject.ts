import { DomainObject } from './DomainObject';

/**
 * In Domain Driven Design, a Value Object is a type of Domain Object for which:
 * - properties are immutable (e.g., its a fact)
 * - identity does not matter
 *   - i.e., it is uniquely identifiable by its properties
 *   - i.e., if you change the value of any of its properties, it is now considered a new value-object
 *
 * For example,
 * - A `Address { street, city, state, country }` is a value object. Changing any of the properties, like `street`, produces a completely new address
 * - A `Name { first, last }` is a value object. Changing either property means you are dealing with a new name.
 */
export abstract class DomainValueObject<T> extends DomainObject<T> {}
