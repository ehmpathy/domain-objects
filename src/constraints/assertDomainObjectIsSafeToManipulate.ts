import { DomainObject } from '../instantiation/DomainObject';

export class DomainObjectNotSafeToManipulateError extends Error {
  constructor({ unsafeKeys, className }: { unsafeKeys: string[]; className: string }) {
    const message = `
DomainObject '${className}' is not safe to manipulate.

Specifically, the keys: ${JSON.stringify(unsafeKeys)}.

Please make sure all nested objects are DomainObjects and are explicitly defined on the class definition, using the 'nested' static property.

For example:
\`\`\`ts
  class ${className} ... {
    public static nested = { ${unsafeKeys[0]}: ... };
  }
\`\`\`
    `;
    super(message);
  }
}

/**
 * For internal use only. DomainObjects must have all nested domain objects defined in their `DomainObject.nested` definition. Otherwise, we could be dealing with a domain object that is not instantiated, and thus we will not appropriately serialize it.
 *
 * Therefore, if any DomainObject has nested objects that are not instantiated as DomainObjects, there is a risk that the nested object is really a nested, uninstantiated DomainObject - which would cause bugs if uncaught.
 */
export const assertDomainObjectIsSafeToManipulate = <T extends Record<string, any>>(obj: DomainObject<T> & Record<string, any>): void => {
  // grab all the keys that have objects defined for their values
  const nestedObjectKeys = Object.keys(obj).filter((key) => typeof obj[key] === 'object' && obj[key] !== null);

  // grab all the "nested object keys" who's values are defined as "should be domain object"
  const nestedDomainObjectKeysDefined = Object.keys((obj.constructor as typeof DomainObject).nested ?? {});
  const nestedNonDomainObjectObjectKeys = nestedObjectKeys.filter((key) => !nestedDomainObjectKeysDefined.includes(key)); // not explicitly defined as a nested domain object key

  // now apply some filters to those concerning keys
  let concerningKeys = nestedNonDomainObjectObjectKeys;

  // now filter out the keys which correspond to well known objects that we known are serializable and are not potentially DomainObjects (e.g., dates)
  concerningKeys = concerningKeys.filter((concerningKey) => {
    const value = obj[concerningKey];
    if (value instanceof Date) return false; // if its a date, not concerning
    if (value instanceof Buffer) return false; // if its a buffer, not concerning
    return true; // if it wasn't filtered out by now, its still concerning
  });

  // now filter out the keys which correspond to arrays of objects which are safe to manipulate; // TODO: consider how we can move this check to not be dependent on runtime contents of the domain object instance
  concerningKeys = concerningKeys.filter((concerningKey) => {
    const value = obj[concerningKey];
    if (!Array.isArray(value)) return true; // if its not an array, its still concerning
    try {
      assertDomainObjectIsSafeToManipulate(value);
      return false; // if we reached here, that means that the contents of the array _are_ safe, so not concerning
    } catch (error) {
      if (error instanceof DomainObjectNotSafeToManipulateError) return true; // if it was this error, that means that this key is definitely still concerning
      throw error; // otherwise its an error we couldn't handle, so we should pass it up
    }
  });

  // if we have any concerningKeys still, then its not safe!
  if (concerningKeys.length) throw new DomainObjectNotSafeToManipulateError({ unsafeKeys: concerningKeys, className: obj.constructor.name });
};
