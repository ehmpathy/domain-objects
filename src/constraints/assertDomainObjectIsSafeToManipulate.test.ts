import { v4 as uuid } from 'uuid';

import { DomainObject } from '../instantiation/DomainObject';
import { assertDomainObjectIsSafeToManipulate } from './assertDomainObjectIsSafeToManipulate';

describe('assertDomainObjectIsSafeToManipulate', () => {
  it('should not throw an error if DomainObject has no nested objects', () => {
    // define the plant pot
    interface PlantPot {
      diameterInInches: number;
    }
    class PlantPot extends DomainObject<PlantPot> implements PlantPot {}

    // instantiate it
    const pot = new PlantPot({ diameterInInches: 7 });

    // check if its safe
    assertDomainObjectIsSafeToManipulate(pot); // should be safe
  });
  it('should not throw an error if DomainObject has a nested Date object', () => {
    // define the plant pot
    interface PlantPot {
      createdAt: Date;
      diameterInInches: number;
    }
    class PlantPot extends DomainObject<PlantPot> implements PlantPot {}

    // instantiate it
    const pot = new PlantPot({ createdAt: new Date(), diameterInInches: 7 });

    // check if its safe
    assertDomainObjectIsSafeToManipulate(pot); // should be safe
  });
  it('should not throw an error if all nested objects on DomainObject are explicitly defined as nested', () => {
    // define the plant pot
    interface PlantPot {
      diameterInInches: number;
    }
    class PlantPot extends DomainObject<PlantPot> implements PlantPot {}

    // define the plant
    interface Plant {
      pot: PlantPot;
      lastWatered: string;
    }
    class Plant extends DomainObject<Plant> implements Plant {
      public static nested = { pot: PlantPot };
    }

    // now show we still think its safe
    const plant = new Plant({
      pot: { diameterInInches: 7 },
      lastWatered: 'monday',
    });
    assertDomainObjectIsSafeToManipulate(plant);
  });
  it('should not throw an error if DomainObject has an array of strings', () => {
    // define the plant pot
    interface PlantPot {
      diameterInInches: number;
      pastPlantUuids: string[];
    }
    class PlantPot extends DomainObject<PlantPot> implements PlantPot {}

    // instantiate it
    const pot = new PlantPot({
      diameterInInches: 7,
      pastPlantUuids: [uuid(), uuid()],
    });

    // check if its safe
    assertDomainObjectIsSafeToManipulate(pot); // should be safe
  });
  it('should not throw an error if DomainObject has an array of DomainObjects explicitly defined as nested', () => {
    // define the plant pot
    interface PlantPot {
      diameterInInches: number;
    }
    class PlantPot extends DomainObject<PlantPot> implements PlantPot {}

    // define the plant
    interface Plant {
      pastPots: PlantPot[];
      lastWatered: string;
    }
    class Plant extends DomainObject<Plant> implements Plant {
      public static nested = { pastPots: PlantPot };
    }

    // now show we still think its safe
    const plant = new Plant({
      pastPots: [{ diameterInInches: 7 }],
      lastWatered: 'monday',
    });
    assertDomainObjectIsSafeToManipulate(plant);
  });
  it('should throw a helpful error if any of the properties are a nested object not explicitly defined', () => {
    // define the plant pot
    interface PlantPot {
      diameterInInches: number;
    }
    class PlantPot extends DomainObject<PlantPot> implements PlantPot {}

    // define the plant
    interface Plant {
      pot: PlantPot;
      lastWatered: string;
    }
    class Plant extends DomainObject<Plant> implements Plant {}

    // now show we do not think its safe
    const plant = new Plant({
      pot: { diameterInInches: 7 },
      lastWatered: 'monday',
    });
    try {
      assertDomainObjectIsSafeToManipulate(plant);
      throw new Error('should not reach here');
    } catch (error) {
      if (!(error instanceof Error)) throw error;
      expect(error.message).toContain(
        `DomainObject 'Plant' is not safe to manipulate.`,
      );
      expect(error.message).toContain(`["pot"]`); // should identify the culprits
      expect(error.message).toContain(
        `Please make sure all nested objects are DomainObjects and are explicitly defined on the class definition, using the 'nested' static property.`,
      ); // should explain how to fix
      expect(error.message).toContain('For example'); // should include an example
      expect(error.message).toMatchSnapshot(); // should look nice - visually
    }
  });
  it('should throw a helpful error if any of the properties are an array with nested objects not explicitly defined', () => {
    // define the plant pot
    interface PlantPot {
      diameterInInches: number;
    }
    class PlantPot extends DomainObject<PlantPot> implements PlantPot {}

    // define the plant
    interface Plant {
      pastPots: PlantPot[];
      lastWatered: string;
    }
    class Plant extends DomainObject<Plant> implements Plant {}

    // now show we do not think its safe
    const plant = new Plant({
      pastPots: [{ diameterInInches: 7 }],
      lastWatered: 'monday',
    });
    try {
      assertDomainObjectIsSafeToManipulate(plant);
      throw new Error('should not reach here');
    } catch (error) {
      if (!(error instanceof Error)) throw error;
      expect(error.message).toContain(
        `DomainObject 'Plant' is not safe to manipulate.`,
      );
      expect(error.message).toContain(`["pastPots"]`); // should identify the culprits
      expect(error.message).toContain(
        `Please make sure all nested objects are DomainObjects and are explicitly defined on the class definition, using the 'nested' static property.`,
      ); // should explain how to fix
      expect(error.message).toContain('For example'); // should include an example
      expect(error.message).toMatchSnapshot(); // should look nice - visually
    }
  });
});
