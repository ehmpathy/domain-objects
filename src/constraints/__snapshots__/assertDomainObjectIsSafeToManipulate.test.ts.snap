// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assertDomainObjectIsSafeToManipulate should throw a helpful error if any of the properties are a nested object not explicitly defined 1`] = `
"
DomainObject 'Plant' is not safe to manipulate.

Specifically, the keys: [\\"pot\\"].

Please make sure all nested objects are DomainObjects and are explicitly defined on the class definition, using the 'nested' static property.

For example:
\`\`\`ts
  class Plant ... {
    public static nested = { pot: ... };
  }
\`\`\`
    "
`;

exports[`assertDomainObjectIsSafeToManipulate should throw a helpful error if any of the properties are an array with nested objects not explicitly defined 1`] = `
"
DomainObject 'Plant' is not safe to manipulate.

Specifically, the keys: [\\"pastPots\\"].

Please make sure all nested objects are DomainObjects and are explicitly defined on the class definition, using the 'nested' static property.

For example:
\`\`\`ts
  class Plant ... {
    public static nested = { pastPots: ... };
  }
\`\`\`
    "
`;
